### PR TITLE
[cmd/telemetrygen] Improvements and fixes to telemetrygen traces

### PR DESCRIPTION
**Pull request by liustanley**
**Description:** <Describe what has changed.>

- Adds new option to disable the batchspanprocessor for testing large trace sizes (enabled by default)
- Adds new option to configure size in MB of each trace generated for load testing
- Fixes issue where large trace payloads (>2000) wouldn't generate the correct number of traces, needed to add delay after each trace is generated

### CODE REVIEW

File: .chloggen/telemetrygen-traces-update.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Adds options to configure the use of the batchspanprocessor and the size in MB of each trace for load testing.
+
+# One or more tracking issues related to the change
+issues: [9597]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Fixes issue where service name option didn't properly set on the trace, and fixes issue where large trace payloads (>2000) wouldn't generate the correct number of traces.
```
Comment (mx-psi) [inline .chloggen/telemetrygen-traces-update.yaml:16]: Instead of listing several fixes here let's split into several changelog entries in separate files

File: cmd/telemetrygen/internal/traces/worker.go [hunk 2]
```diff
@@ -69,14 +78,15 @@ func (w worker) simulateTraces() {
...
-			semconv.PeerServiceKey.String("telemetrygen-client"),
+			semconv.PeerServiceKey.String(w.serviceName),
 		))
 
 		if err := limiter.Wait(context.Background()); err != nil {
 			w.logger.Fatal("limiter waited failed, retry", zap.Error(err))
 		}
 
 		opt := trace.WithTimestamp(time.Now().Add(fakeSpanDuration))
+		time.Sleep(time.Duration(rand.Intn(40)) * time.Millisecond)
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/traces/worker.go:89]: ditto, I don't get this, why is this needed? I don't think we should sleep if we can avoid it

File: cmd/telemetrygen/internal/traces/worker.go [hunk 3]
```diff
@@ -53,8 +59,11 @@ func (w worker) simulateTraces() {
 		ctx, sp := tracer.Start(context.Background(), "lets-go", trace.WithAttributes(
 			attribute.String("span.kind", "client"), // is there a semantic convention for this?
 			semconv.NetPeerIPKey.String(fakeIP),
-			semconv.PeerServiceKey.String("telemetrygen-server"),
+			semconv.PeerServiceKey.String(w.serviceName),
 		))
+		for j := 0; j < w.loadSize; j++ {
+			sp.SetAttributes(attribute.String(fmt.Sprintf("load-%v", j), string(make([]byte, charactersPerMB))))
```
[the remark this answers is not in the document]
Comment (liustanley) [reply cmd/telemetrygen/internal/traces/worker.go:65]: This attribute is so that we can test traces with large payloads, which is something we needed in testing the max receive size of the OTLP receiver gRPC endpoint. For example, we can set loadSize to 3 and it will generate a trace of size 3 MB which will be accepted by the receiver. Or we can set loadSize to 5 and it will generate a trace of size 5 MB which will not be accepted by the receiver with the default max size.

File: cmd/telemetrygen/internal/traces/worker.go (hunk 2 above)
Comment (codeboten) [reply cmd/telemetrygen/internal/traces/worker.go:89]: This seems like a bug somewhere else

Comment (mx-psi) [reply cmd/telemetrygen/internal/traces/worker.go:89]: > This seems like a bug somewhere else

Agreed. I am okay with the other change (or something like it), but I don't think we should fix this with the `time.Sleep` approach, let's keep the PR only for the `loadSize` option and open an issue to discuss this bug and how to solve it.

File: .chloggen/telemetrygen-traces-fix-service-name-and-large-payloads.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: telemetrygen
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fixes issue where service name option didn't properly set on the trace, and fixes issue where large trace payloads (>2000) wouldn't generate the correct number of traces.
```
Comment (mx-psi) [inline .chloggen/telemetrygen-traces-fix-service-name-and-large-payloads.yaml:8]: Let's remove the large payloads mention and revert these changes

```suggestion
note: `telemetrygen traces` now correctly sets the `service.name` resource attribute on generated traces
```

File: .chloggen/telemetrygen-traces-batch-size-options.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: telemetrygen
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Adds options to configure the use of the batchspanprocessor and the size in MB of each trace for load testing.
```
Comment (mx-psi) [inline .chloggen/telemetrygen-traces-batch-size-options.yaml:8]: Could you name the options here? I would also talk about batching without making reference to the specific implementation using the batchspanprocessor

File: cmd/telemetrygen/internal/traces/config.go [hunk 6]
```diff
@@ -34,4 +36,6 @@ func (c *Config) Flags(fs *pflag.FlagSet) {
 	fs.IntVar(&c.NumTraces, "traces", 1, "Number of traces to generate in each worker (ignored if duration is provided)")
 	fs.BoolVar(&c.PropagateContext, "marshal", false, "Whether to marshal trace context via HTTP headers")
 	fs.StringVar(&c.ServiceName, "service", "telemetrygen", "Service name to use")
+	fs.BoolVar(&c.Batch, "batch", true, "Whether to use batch span processor")
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/traces/config.go:39]:
```suggestion
	fs.BoolVar(&c.Batch, "batch", true, "Whether to batch traces")
```

File: cmd/telemetrygen/internal/traces/config.go [hunk 7]
```diff
@@ -34,4 +36,6 @@ func (c *Config) Flags(fs *pflag.FlagSet) {
...
+	fs.IntVar(&c.LoadSize, "size", 0, "Desired size in MB for each trace generated")
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/traces/config.go:40]: With the current implementation this sounds like the 'minimum size' instead of the exact size. Am I right?

File: cmd/telemetrygen/internal/traces/worker.go [hunk 8]
```diff
@@ -37,12 +39,16 @@ type worker struct {
...
+	serviceName      string
+	loadSize         int
 }
 
 const (
 	fakeIP string = "1.2.3.4"
 
 	fakeSpanDuration = 123 * time.Microsecond
+
+	charactersPerMB = 1048577
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/traces/worker.go:51]: Can you describe this variable and how you arrived at the value?

File: cmd/telemetrygen/internal/traces/worker.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mx-psi) [reply cmd/telemetrygen/internal/traces/worker.go:65]: Since the user can see this attribute, I would add this to the description of the `size` option.

File: cmd/telemetrygen/internal/traces/worker_test.go [hunk 9]
```diff
@@ -92,7 +92,7 @@ func TestUnthrottled(t *testing.T) {
 
 	cfg := &Config{
 		Config: common.Config{
-			TotalDuration: 50 * time.Millisecond,
+			TotalDuration: 50 * time.Millisecond * 40,
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/traces/worker_test.go:95]: I think we can revert this since we are not adding the sleep on this PR

**Review by mx-psi — DISMISSED** [review]
Per https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21629#discussion_r1195021732, let's remove the `time.Sleep` and associated changelog entry and unit test changes and focus the PR on the other improvements

File: cmd/telemetrygen/internal/traces/worker.go (hunk 2 above)
Comment (liustanley) [reply cmd/telemetrygen/internal/traces/worker.go:89]: Just opened an issue here: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22827

**Review by codeboten — COMMENTED** [review]
@mx-psi ptal as you have a request for change

**Comment by mx-psi** [comment]
Sorry for the long delay. I have not further comments other than adressing @codeboten's question

File: cmd/telemetrygen/internal/traces/worker.go [hunk 10]
```diff
@@ -53,8 +58,11 @@ func (w worker) simulateTraces() {
 		ctx, sp := tracer.Start(context.Background(), "lets-go", trace.WithAttributes(
 			attribute.String("span.kind", "client"), // is there a semantic convention for this?
 			semconv.NetPeerIPKey.String(fakeIP),
-			semconv.PeerServiceKey.String("telemetrygen-server"),
+			semconv.PeerServiceKey.String(w.serviceName),
```
[the remark this answers is not in the document]
Comment (liustanley) [reply cmd/telemetrygen/internal/traces/worker.go]: I tested again and this change is no longer needed, maybe something changed on Datadog's end but the service name is setting correctly now.

**Review by mx-psi — APPROVED** [review]
